src/datumaro/plugins/configurable_validator.py

```diff
@@ -0,0 +1,732 @@
+                        else:
+                            self.stats[label_name]["attributes"][attr][str(value)] = 1
+                    except Exception as e:
+                        log.warning(
+                            "Label '%s': failed to get access to attribute %s" % (label_name, e)
+                        )
+                        if {UndefinedAttribute} & self.warnings:
+                            # attribute is not defined within the label
+                            self.stats["undefined_attribute"].add(item_key + (label_name, attr))
+
+                if {MissingAttribute} & self.warnings:
+                    # defined attribute is not found in this annotation
+                    for attr in self.stats[label_name]["attributes"]:
+                        if attr not in annotation.attributes:
+                            self.stats["missing_attribute"].add(item_key + (label_name, attr))
+        else:
+            label_name = str(annotation.label)
+            if {UndefinedLabel} & self.warnings:
+                self.stats["undefined_label"].add(item_key + (label_name,))
+
+            if {UndefinedAttribute} & self.warnings:
+                for attr, value in annotation.attributes.items():
+                    self.stats["undefined_attribute"].add(item_key + (label_name, attr))
+
+    def _update_item_type_stats(self, item: DatasetItem):
+        NotImplemented
+
+    def _update_ann_type_stats(self, item_key: tuple, annotation: Annotation):
+        NotImplemented
+
+
+class ClsStats(_BaseAnnStats):
+    def __init__(self, label_categories: LabelCategories, warnings: set):
+        super().__init__(label_categories=label_categories, warnings=warnings)
+
+        self.stats["multiple_label"] = set()
+
+    def _update_item_type_stats(self, item: DatasetItem):
+        item_key = (item.id, item.subset)
+
+        if {MultiLabelAnnotations} & self.warnings:
+            if sum([1 for ann in item.annotations if ann.type == AnnotationType.label]) > 1:
+                self.stats["multiple_label"].add(item_key)
+
+    def _update_ann_type_stats(self, item_key: tuple, annotation: Annotation):
+        pass
+
+
+class DetStats(_BaseAnnStats):
+    BBOX_WARNINGS = {
+        FarFromAttrMean,  # annotation level
+        FarFromLabelMean,  # annotation level
+        ImbalancedDistInAttribute,  # annotation level: bbox
+        ImbalancedDistInLabel,  # annotation level: bbox
+        InvalidValue,  # annotation level
+        NegativeLength,  # annotation level
+    }
+
+    def __init__(self, label_categories: LabelCategories, warnings: set):
+        super().__init__(label_categories=label_categories, warnings=warnings)
+
+        self.items = {}
+        for label_cat in self.label_categories.items:
+            label_name = label_cat.name
+            self.stats[label_name]["pts"] = []  # (id, subset, ann_id, w, h)
+
+        self.stats["invalid_value"] = set()
+        self.stats["negative_length"] = set()
```

Comment:
Could we have a typed schema, e.g. `@dataclass`, for the output data store? It currently uses a dictionary with string key and it is hard to see what is going on this class.

Reply:
I introduced StatsData at cc10ff43ab54b743fc524f29ad1f4bc016eaa04b.